### ISSUE TITLE
Questions on the output in the log

### ISSUE THREAD

**Issue by flippercy** [microsoft/FLAML#31]
Hi Dr. Wang:

Got a few questions from my team on the content in the log of FLAML.

This is part of the log from one of our tests on FLAML (all the numbers on loss are redacted for compliance reasons):

{"record_id": 0, "iter_per_learner": 1, "logged_metric": false, "trial_time": 1756.8860552310944, "total_search_time": 2590.4430527687073, "validation_loss": XXX, "config": {"max_depth": 6, "n_estimators": 100, "min_child_weight": 10, "subsample": 0.67, "colsample_bylevel": 0.9, "gamma": 0, "learning_rate": 0.07435893300587489}, "best_validation_loss": XXX, "best_config": {"max_depth": 6, "n_estimators": 100, "min_child_weight": 10, "subsample": 0.67, "colsample_bylevel": 0.9, "gamma": 0, "learning_rate": 0.07435893300587489}, "learner": "MonotonicXgboostGBTree", "sample_size": 784536}

{"record_id": 1, "iter_per_learner": 5, "logged_metric": false, "trial_time": 1537.3765320777893, "total_search_time": 13424.922722578049, "validation_loss": XXX, "config": {"max_depth": 4, "n_estimators": 110, "min_child_weight": 1, "subsample": 0.5954399576961257, "colsample_bylevel": 1.0, "gamma": 1e-14, "learning_rate": 0.10828032871243709}, "best_validation_loss": XXX "best_config": {"max_depth": 4, "n_estimators": 110, "min_child_weight": 1, "subsample": 0.5954399576961257, "colsample_bylevel": 1.0, "gamma": 1e-14, "learning_rate": 0.10828032871243709}, "learner": "MonotonicXgboostGBTree", "sample_size": 784536}

{"record_id": 2, "iter_per_learner": 13, "logged_metric": false, "trial_time": 340.0606036186218, "total_search_time": 34851.914006233215, "validation_loss": XXX, "config": {"max_depth": 5, "num_leaves": 23, "n_estimators": 157, "min_child_weight": 1, "subsample": 0.5112583180636173, "colsample_bylevel": 0.9863382485941592, "min_split_gain": 1e-14, "learning_rate": 0.05875161500234584}, "best_validation_loss": XXX, "best_config": {"max_depth": 5, "num_leaves": 23, "n_estimators": 157, "min_child_weight": 1, "subsample": 0.5112583180636173, "colsample_bylevel": 0.9863382485941592, "min_split_gain": 1e-14, "learning_rate": 0.05875161500234584}, "learner": "MonotonicLightGBMGBDT", "sample_size": 784536}

{"record_id": 3, "iter_per_learner": 18, "logged_metric": false, "trial_time": 270.2408003807068, "total_search_time": 41368.91024374962, "validation_loss": XXX, "config": {"max_depth": 4, "n_estimators": 338, "min_data_in_leaf": 56, "subsample": 0.6614322871324126, "colsample_bylevel": 0.9458919560564311, "learning_rate": 0.23062756268773424}, "best_validation_loss": XXX, "best_config": {"max_depth": 4, "n_estimators": 338, "min_data_in_leaf": 56, "subsample": 0.6614322871324126, "colsample_bylevel": 0.9458919560564311, "learning_rate": 0.23062756268773424}, "learner": "MonotonicCatboost", "sample_size": 784536}

{"record_id": 4, "iter_per_learner": 22, "logged_metric": false, "trial_time": 366.1694631576538, "total_search_time": 43080.46155285835, "validation_loss": XXX, "config": {"max_depth": 4, "n_estimators": 448, "min_data_in_leaf": 46, "subsample": 0.6950654501710251, "colsample_bylevel": 0.956150967914549, "learning_rate": 0.4527543463119874}, "best_validation_loss": XXX, "best_config": {"max_depth": 4, "n_estimators": 448, "min_data_in_leaf": 46, "subsample": 0.6950654501710251, "colsample_bylevel": 0.956150967914549, "learning_rate": 0.4527543463119874}, "learner": "MonotonicCatboost", "sample_size": 784536}

{"record_id": 5, "iter_per_learner": 23, "logged_metric": false, "trial_time": 343.4558777809143, "total_search_time": 45475.49441862106, "validation_loss": XXX, "config": {"max_depth": 4, "n_estimators": 405, "min_data_in_leaf": 58, "subsample": 0.734450014003538, "colsample_bylevel": 0.9644762947991873, "learning_rate": 0.3151376812002405}, "best_validation_loss": XXX, "best_config": {"max_depth": 4, "n_estimators": 405, "min_data_in_leaf": 58, "subsample": 0.734450014003538, "colsample_bylevel": 0.9644762947991873, "learning_rate": 0.3151376812002405}, "learner": "MonotonicCatboost", "sample_size": 784536}

I am wondering:

1. What does 'iter_per_learner' mean? My understanding is that the output in the log was generated in batch. For example, for record_id 2, does it include 13 or 8 (13-5 from record_id 1) MonotonicLightGBMGBDT models with different sets of hyperparameters?

2. What does 'trial_time' mean? How is it different from 'total_search_time"?

3. What is the difference between 'config' and 'best_config' in each record? They all look the same.

3. If the process reaches the time budget in the middle of an iteration, will it stop immediately or finish the current iteration first before stopping?

Appreciate your help! As you can see from the log, our dataset is quite large (780000+ records and thousands of predictors). Although the fitting is far from over yet, the current optimal result is already as good as what we got using BayesOpt.

Best,

**Comment by sonichi** [comment]
Thanks for sharing the feedback about flaml in action. Great to hear that its result is good in your task. To the questions:

1. `iter_per_learner` means how many models have been tried for each learner. The reason you see `iter_per_learner=13` for `record_id=2` is that flaml only logs better configs than the previous iters by default, i.e., `log_type='better'`. 
2. `trial_time` means the time taken to train and evaluate one config in that trial. `total_search_time` is the total time spent from the beginning of `fit()`.
3. Because `log_type='better'` by default, `config` is equal to `best_config`. If you use `log_type='all'` instead, all the trials will be logged. And then `config` corresponds to the config in that iteration, and `best_config` is the best config so far.
4. flaml will adjust the `n_estimators` for lightgbm etc. according to the remaining budget and check the time budget constraint and stop in several places. Most of the time that makes `fit()` stops before the given budget. Occasionally it may run over the time budget slightly. But the log file always contains the best config info and you can recover it using `retrain_from_log()`.

**Comment by flippercy** [comment]
Thank you! All make sense.

**Comment by sonichi** [comment]
Great. If flaml is useful in your application, it will be nice to have your feedback. Feel free to email hpo@microsoft.com to ask questions or share your feedback.
